plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php

```diff
@@ -671,4 +674,181 @@ private function normalize_raw_meta_data( array $data ) : array {
+			$assoc_args,
+			array(
+				'for-new-shop' => false,
+				'with-sync'    => false,
+			)
+		);
+
+		$enable_hpos = true;
+		WP_CLI::log( __( 'Running pre-enable checks...', 'woocommerce' ) );
+
+		$is_new_shop = \WC_Install::is_new_install();
+		if ( $assoc_args['for-new-shop'] && ! $is_new_shop ) {
+			WP_CLI::warning( __( '[Failed] This is not a new shop, but --for-new-shop flag was passed.', 'woocommerce' ) );
+			$enable_hpos = false;
+		}
+
+		/** Feature controller instance @var FeaturesController $feature_controller */
+		$feature_controller = wc_get_container()->get( FeaturesController::class );
+		$plugin_info        = $feature_controller->get_compatible_plugins_for_feature( 'custom_order_tables', true );
+		if ( count( array_merge( $plugin_info['uncertain'], $plugin_info['incompatible'] ) ) > 0 ) {
+			WP_CLI::warning( __( '[Failed] Some installed plugins are incompatible. Please review the plugins by going to WooCommerce > Settings > Advanced > Features and see the Data storage for orders section.', 'woocommerce' ) );
+			$enable_hpos = false;
+		}
+
+		/** DataSynchronizer instance @var DataSynchronizer $data_synchronizer */
+		$data_synchronizer = wc_get_container()->get( DataSynchronizer::class );
+		$pending_orders    = $data_synchronizer->get_total_pending_count();
+		$table_exists      = $data_synchronizer->check_orders_table_exists();
+
+		if ( ! $table_exists ) {
+			WP_CLI::warning( __( 'Orders table does not exist. Creating...', 'woocommerce' ) );
+			if ( $is_new_shop || 0 === $pending_orders ) {
+				$data_synchronizer->create_database_tables();
+				if ( $data_synchronizer->check_orders_table_exists() ) {
+					WP_CLI::log( __( 'Orders table created.', 'woocommerce' ) );
+					$table_exists = true;
+				} else {
+					WP_CLI::warning( __( '[Failed] Orders table could not be created.', 'woocommerce' ) );
+					$enable_hpos = false;
+				}
+			} else {
+				WP_CLI::warning( __( '[Failed] The orders table does not exist and this is not a new shop. Please create the table by going to WooCommerce > Settings > Advanced > Features and enabling sync.', 'woocommerce' ) );
+				$enable_hpos = false;
+			}
+		}
+
+		if ( $pending_orders > 0 ) {
+			WP_CLI::warning(
+				sprintf(
+					// translators: %s is the command to run (wp wc cot sync).
+					__( '[Failed] There are orders pending sync. Please run `%s` to sync pending orders.', 'woocommerce' ),
+					'wp wc cot sync',
+				)
+			);
+			$enable_hpos = false;
+		}
+
+		if ( $assoc_args['with-sync'] && $table_exists ) {
+			if ( $data_synchronizer->data_sync_is_enabled() ) {
+				WP_CLI::warning( __( 'Sync is already enabled.', 'woocommerce' ) );
+			} else {
+				$feature_controller->change_feature_enable( DataSynchronizer::ORDERS_DATA_SYNC_ENABLED_OPTION, true );
+				WP_CLI::log( __( 'Sync enabled.', 'woocommerce' ) );
+			}
+		}
+
+		if ( ! $enable_hpos ) {
+			WP_CLI::error( __( 'HPOS could not be enabled, see the errors above', 'woocommerce' ) );
+			return;
+		}
```

Comment:
(2/2) Then this will also be met. That's fine but, in cases where HPOS is already enabled, the output presented to the user will look like this:

```
Running pre-enable checks...
Warning: [Failed] Some installed plugins are incompatible. Please review the plugins by going to WooCommerce > Settings > Advanced > Features and see the Data storage for orders section.
Error: HPOS could not be enabled, see the errors above
```

It's not wrong, but it's arguably misleading and it would imply to me that HPOS is not enabled. In this scenario, would it be better if we bailed with:

```
Warning: Some installed plugins are incompatible. Please review the plugins by going to WooCommerce > Settings > Advanced > Features and see the Data storage for orders section.
Error: HPOS is already enabled.
```

(That is, still warn about the plugin and also let them know HPOS is enabled.)

Reply:
I have changed the language to say that pre-checks failed, without saying anything about whether HPOS is enabled or not. Perhaps there is a utility in a `wp wc cot status` command, but maybe in a different PR.